lib/chef/provider/aws_cloudsearch_domain.rb

```diff
@@ -0,0 +1,126 @@
+require 'chef/provisioning/aws_driver/aws_provider'
+
+class Chef::Provider::AwsCloudsearchDomain < Chef::Provisioning::AWSDriver::AWSProvider
+
+  def create_aws_object
+    converge_by "Creating new CloudSearch domain #{new_resource.domain_name}" do
+      create_domain
+    end
+    # multi_az defaults to false on creation
+    # so we only need to hit the API if it is true
+    if new_resource.multi_az
+      converge_by "Setting availability options for CloudSearch domain #{new_resource}" do
+        update_availability_options
+      end
+    end
+
+    if update_scaling_params?
+      converge_by "Setting scaling parameters for CloudSearch domain #{new_resource.domain_name}" do
+        update_scaling_parameters
+      end
+    end
+
+    if update_policy?
+      converge_by "Setting access policy for CloudSearch domain #{new_resource.domain_name}" do
+        update_service_access_policy
+      end
+    end
+
+    if update_index_fields?
+      converge_by "Defining index fields for CloudSearch domain #{new_resource.domain_name}" do
+        new_resource.index_fields.each do |idx_field|
+          if !idx_field.is_a? Hash
+            raise ArgumentError, "Expected index_fields to contain only Hashes"
+          else
+            create_index_field(idx_field)
+          end
+        end
+      end
+    end
+  end
+
+  def destroy_aws_object(instance)
+    converge_by "Deleting CloudSearch domain #{new_resource.domain_name}" do
+      cs_client.delete_domain(domain_name: new_resource.domain_name)
+    end
+  end
+
+  def update_aws_object(instance)
+    if update_availability_options?
+      converge_by "Updating availability options for CloudSearch domain #{new_resource}" do
+        update_availability_options
+      end
+    end
+
+    if update_scaling_params?
+      converge_by "Updating scaling parameters for CloudSearch domain #{new_resource.domain_name}" do
+        update_scaling_parameters
+      end
+    end
+
+    if update_policy?
+      converge_by "Updating access policy for CloudSearch domain #{new_resource.domain_name}" do
+        update_service_access_policy
+      end
+    end
+
+    if update_index_fields?
+      Chef::Log.warn("Updating existing index_fields not currently supported")
+    end
+  end
+
+
+  # TODO(ssd): Expand these to be idempotent?  It takes an extra API
```

Comment:
I started in on making this idempotent and remembered what I meant by "extra API call".  The describe_domain api call that we use to get the object doesn't return availability, scalability, indexing, or access policy info.  Each piece of that info is a separate call:

```
irb(main):023:0> pp AWS::CloudSearch::Client::V20130101.new.describe_domains(domain_names: ["foo"])[:domain_status_list].first; nil
{:created=>true,
 :requires_index_documents=>false,
 :domain_id=>"862552916454/foo",
 :doc_service=>nil,
 :search_partition_count=>0,
 :deleted=>false,
 :search_service=>nil,
 :domain_name=>"foo",
 :processing=>true,
 :search_instance_count=>0,
 :arn=>"arn:aws:cloudsearch:us-east-1:862552916454:domain/foo",
 :limits=>
  {:available_instance_types=>{:member=>"search.m3.2xlarge"},
   :maximum_replication_count=>5,
   :maximum_partition_count=>10}}
irb(main):024:0> pp AWS::CloudSearch::Client::V20130101.new.describe_availability_options(domain_name: "foo")
{:availability_options=>{:options=>false, :status=>{:update_date=>2015-08-04 23:27:05 UTC, :state=>"Processing", :creation_date=>2015-08-04 23:27:05 UTC, :update_version=>2, :pending_deletion=>false}}, :response_metadata=>{:request_id=>"9b8826fc-3b03-11e5-90cc-a1231d3e2650"}}
```